File: Solution-Go/Palindrome_234.go
func isPalindrome(head *ListNode) bool {
    endOfFirstHalf := getEndOfFirstHalf(head)
    startOfSecondHalfReversed := reverseLinkedList(endOfFirstHalf.Next)
    tail := startOfSecondHalfReversed
    for tail != nil {
        if tail.Val != head.Val {
            return false
        }
        tail = tail.Next
        head = head.Next
    }
    reverseLinkedList(startOfSecondHalfReversed)
    return true
}

func getEndOfFirstHalf(head *ListNode) *ListNode {
    fast, slow := head, head
    for fast.Next != nil && fast.Next.Next != nil {
        fast = fast.Next.Next
        slow = slow.Next
    }
    return slow
}

func reverseLinkedList(head *ListNode) (newHead *ListNode) {
    for head != nil {
        next := head.Next
        head.Next = newHead
        newHead = head
        head = next
    }
    return
}

File: Solution-Python/Palindrome_234.py
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        endOfFirstHalf = self.getEndOfFirstHalf(head)
        startOfSecondHalfReversed = self.reverseLinkedList(endOfFirstHalf.next)
        tail = startOfSecondHalfReversed
        while tail:
            if tail.val != head.val:
                return False
            tail = tail.next
            head = head.next
        self.reverseLinkedList(startOfSecondHalfReversed)
        return True
        
    def getEndOfFirstHalf(self, head: ListNode) -> ListNode:
        fast, slow = head, head
        while fast.next and fast.next.next:
            fast = fast.next.next
            slow = slow.next
        return slow
    
    def reverseLinkedList(self, head: ListNode):
        newHead = None
        while head:
            temp = head.next
            head.next = newHead
            newHead = head
            head = temp
        return newHead